### PR TITLE
Encrypt network communication using the noise protocol

### LINKED ISSUE #229: Introduce authentication and encryption for the network layer
Eventually we will transition the network layer to libp2p so we have a more structured way of designing network communication.

Up until then, we still want proper identification of peers (our current `TakerId` is transient with every connection), authentication of each other's messages and encryption of the entire communication.

To make the transition to libp2p later on easier, we first will:

- Use [`snow`](https://docs.rs/snow/0.8.0/snow/index.html) for the heavy lifting of encrypting and authenticating the connection
- Use the libp2p multi-addresses specification to configure the connection details to the other party

Using snow will require that we know the public key of the party ahead of time. Using libp2p's multiaddress provides several benefits:

- Specify connection details + public key in a single argument
- Allows usage of domains as well as IP addresses (https://github.com/comit-network/hermes/issues/292)

### PR DESCRIPTION
**Pull request by rishflab**
The messages which contain CET's do not fit into the
65535 byte message size limit of the noise protocol.
These messages are chunked to allow compatibility
with noise.

More research is required into the NOISE_PARAMS and
whether they can be used to configure snow elegantly
handle authentication.

### CODE REVIEW

File: daemon/src/noise.rs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    Ok(noise)
+}
+
+pub async fn responder_handshake(connection: &mut TcpStream) -> Result<TransportState> {
+    let builder: Builder<'_> = Builder::new(NOISE_PARAMS.parse()?);
+    let static_key = builder.generate_keypair()?.private;
+
+    let mut noise = builder.local_private_key(&static_key).build_responder()?;
+
+    let mut buf = vec![0u8; 65535];
```
Comment (klochowicz) [inline daemon/src/noise.rs]: nit: `s/65535/NOISE_MAX_MSG_LEN` ?

**Review by klochowicz — APPROVED** [review]
great job solving the large message problem! 

I'm happy for this to go in if @da-kami 's tests pass with it :shipit: 

side note: it probably would have been easier to review if it were 2 commits, the original that already had the green light and then "Support large messages by chunking encryption" to see the difference.

**Comment by da-kami** [comment]
would be cool if we can test rollover successfully as well

**Comment by rishflab** [comment]
> would be cool if we can test rollover successfully as well

I did a manual test and I think it works

```
2021-10-30 10:49:19  INFO GET /api/feed text/event-stream:    
2021-10-30 10:49:21  INFO POST /api/order/sell application/json:    
2021-10-30 10:49:21  INFO POST /api/calculate/margin text/plain:    
2021-10-30 10:49:21  INFO POST /api/calculate/margin text/plain:    
2021-10-30 10:49:29  INFO POST /api/cfd/41deaa9c-d69e-4414-8090-356801997e36/rollOver:    
2021-10-30 10:49:29  INFO Received proposal from the taker 88a932c0-ed62-43f0-8b50-167d69573620: RollOverProposal { order_id: OrderId(Hyphenated(41deaa9c-d69e-4414-8090-356801997e36)), timestamp: Timestamp(1635551369) } to roll over order 41deaa9c-d69e-4414-8090-356801997e36
2021-10-30 10:49:39  INFO POST /api/cfd/41deaa9c-d69e-4414-8090-356801997e36/acceptRollOver:    
2021-10-30 10:49:39 DEBUG Maker accepts a roll_over proposal order_id=41deaa9c-d69e-4414-8090-356801997e36
2021-10-30 10:49:39  INFO Roll; over request got accepted order_id=41deaa9c-d69e-4414-8090-356801997e36
```

**Comment by da-kami** [comment]
> > would be cool if we can test rollover successfully as well
> 
> I did a manual test and I think it works
> 
> ```
> 2021-10-30 10:49:19  INFO GET /api/feed text/event-stream:    
> 2021-10-30 10:49:21  INFO POST /api/order/sell application/json:    
> 2021-10-30 10:49:21  INFO POST /api/calculate/margin text/plain:    
> 2021-10-30 10:49:21  INFO POST /api/calculate/margin text/plain:    
> 2021-10-30 10:49:29  INFO POST /api/cfd/41deaa9c-d69e-4414-8090-356801997e36/rollOver:    
> 2021-10-30 10:49:29  INFO Received proposal from the taker 88a932c0-ed62-43f0-8b50-167d69573620: RollOverProposal { order_id: OrderId(Hyphenated(41deaa9c-d69e-4414-8090-356801997e36)), timestamp: Timestamp(1635551369) } to roll over order 41deaa9c-d69e-4414-8090-356801997e36
> 2021-10-30 10:49:39  INFO POST /api/cfd/41deaa9c-d69e-4414-8090-356801997e36/acceptRollOver:    
> 2021-10-30 10:49:39 DEBUG Maker accepts a roll_over proposal order_id=41deaa9c-d69e-4414-8090-356801997e36
> 2021-10-30 10:49:39  INFO Roll; over request got accepted order_id=41deaa9c-d69e-4414-8090-356801997e36
> ```

I think you should also see the messages where the information was exchanged, similar to the contract setup - are those present in the logs? 

> `2021-10-30 10:49:39  INFO Roll; over request got accepted order_id=41deaa9c-d69e-4414-8090-356801997e36`

This message is printed when rollover is initiated but does not tell you anything about the messages being exchanged successfully.

**Comment by da-kami** [comment]
@rishflab it would  be sweet if we could add an actor test that covers exchanging the collarborative close messages as part of this PR! (i.e.assert that collaborative close was set up)
That would mean a bit more work, but given the currently already existing actor tests it should be relatively trivial to add it. 
This would generally go in the right direction, I think future PRs will have to think about adding actor tests as well when necessary - we cannot just cover those tests separately. 
What do you think @klochowicz ?

**Comment by rishflab** [comment]
> @rishflab it would be sweet if we could add an actor test that covers exchanging the collarborative close messages as part of this PR! (i.e.assert that collaborative close was set up) That would mean a bit more work, but given the currently already existing actor tests it should be relatively trivial to add it. This would generally go in the right direction, I think future PRs will have to think about adding actor tests as well when necessary - we cannot just cover those tests separately. What do you think @klochowicz ?

yeh sure, I can give it a go today. I might need a hand though.

**Comment by klochowicz** [comment]
> @rishflab it would be sweet if we could add an actor test that covers exchanging the collarborative close messages as part of this PR! (i.e.assert that collaborative close was set up) That would mean a bit more work, but given the currently already existing actor tests it should be relatively trivial to add it. This would generally go in the right direction, I think future PRs will have to think about adding actor tests as well when necessary - we cannot just cover those tests separately. What do you think @klochowicz ?

I agree it's a good idea to add the code coverage there, but I'd personally choose to rather do it as a follow-up PR - this one got approved already, and it significantly changes its scope.

Collab close test is something that's also generally useful, not just for the purpose of adding encryption.
Happy to help with any questions you have @rishflab !

**Pull request merged**